packages/block-library/src/navigation-menu-item/edit.js

```diff
@@ -49,41 +45,84 @@ function NavigationMenuItemEdit( {
 	isParentOfSelectedBlock,
 	setAttributes,
 	insertMenuItemBlock,
+	fetchSearchSuggestions,
 } ) {
+	const { label, opensInNewTab, title, url } = attributes;
+	const link = title ? { title, url } : null;
 	const [ isLinkOpen, setIsLinkOpen ] = useState( false );
-	const [ isEditingLink, setIsEditingLink ] = useState( false );
-	const [ urlInput, setUrlInput ] = useState( null );
+	const [ wasClosedByLinkControl, setWasClosedByLinkControl ] = useState( false );
+
+	/**
+	 * It's a kind of hack to handle closing the LinkControl popover
+	 * clicking on the ToolbarButton link.
+	 */
+	useEffect( () => {
+		if ( ! isSelected ) {
+			setIsLinkOpen( false );
+			setWasClosedByLinkControl( false );
+		}
+		return () => {
+			setIsLinkOpen( false );
+			setWasClosedByLinkControl( false );
+		};
+	}, [ isSelected ] );
+
+	// Set the menu item when it's new.
+	// - Open LinkControl popover.
+	useEffect( () => {
+		if ( ! label && isSelected ) setIsLinkOpen( true );
+	}, [] );
 
-	const inputValue = urlInput !== null ? urlInput : url;
+	/**
+	 * `onKeyDown` LinkControl handler.
+	 * It takes over to stop the event propagation to make the
+	 * navigation work, avoiding undesired behaviors.
+	 * For instance, it will block to move between menu items
+	 * when the LinkOver is focused.
+	 *
+	 * @param {Event} event
+	 */
+	const handleLinkControlOnKeyDown = ( event ) => {
+		const { keyCode } = event;
 
-	const onKeyDown = ( event ) => {
-		if ( [ LEFT, DOWN, RIGHT, UP, BACKSPACE, ENTER ].indexOf( event.keyCode ) > -1 ) {
+		if ( [ LEFT, DOWN, RIGHT, UP, BACKSPACE, ENTER ].indexOf( keyCode ) > -1 ) {
 			// Stop the key event from propagating up to ObserveTyping.startTypingInTextField.
 			event.stopPropagation();
 		}
 	};
 
-	const closeURLPopover = () => {
-		setIsEditingLink( false );
-		setUrlInput( null );
-		setIsLinkOpen( false );
-	};
+	/**
+	 * Updates the link attribute when it changes
+	 * through of the `onLinkChange` LinkControl callback.
+	 *
+	 * @param {Object|null} itemLink Link object if it has been selected. Otherwise, Null.
+	 */
+	const updateLink = ( { title: newTitle = '', url: newURL = '' } = {} ) => {
+		setAttributes( {
+			title: newTitle,
+			url: newURL,
+		} );
 
-	const autocompleteRef = useRef( null );
-
-	const onFocusOutside = ( event ) => {
-		const autocompleteElement = autocompleteRef.current;
-		if ( autocompleteElement && autocompleteElement.contains( event.target ) ) {
-			return;
+		// Set the item label as well if it isn't already defined.
+		if ( ! label ) {
+			setAttributes( { label: newTitle } );
 		}
-		closeURLPopover();
 	};
 
-	const stopPropagation = ( event ) => {
-		event.stopPropagation();
+	/**
+	 * It updates the link attribute when the
+	 * link settings changes.
+	 *
+	 * @param {string} setting Setting type, for instance, `new-tab`.
+	 * @param {string} value Setting type value.
+	 */
+	const updateLinkSetting = ( setting, value ) => {
```

Comment:
Can we wrap this with useCallback? Otherwise, we will be generating a new function on each render and given that react assumes we are passing a new prop to the children it may be rerendering them without being necessary.

Reply:
Ok, going to learn about` useCallback()`, but I wonder if the following could work as well:

```es6
const updateLinkSetting = ( setter ) => ( setting, value ) => {
	if ( setting === 'new-tab' ) {
		setter( { opensInNewTab: value } );
	}
};

function NavigationMenuItemEdit( props ) {
	// ...
	return(
		<LinkControl
			onSettingsChange={ updateLinkSetting( setAttribute ) }
		/>
	);
}
```